parser/parser.go

```diff
@@ -2,169 +2,209 @@ package parser
-	m := message.ParamsIncludingMessage{}
-	err := json.Unmarshal(params, &m)
-	if err != nil {
-		return m, lib.ErrEncodingFault
+	if !strings.HasPrefix(m.Channel, "/root") {
+		log.Printf("channel id doesn't start with /root but is %v", m.Channel)
+		return m, lib.ErrRequestDataInvalid
 	}
-	d, err := lib.Decode(m.Channel)
-	m.Channel = string(d)
 	return m, err
 }
 
+// ParseMessage parses a json.RawMessage into a message.Message structure.
 func ParseMessage(msg json.RawMessage) (message.Message, error) {
 	m := message.Message{}
 	err := json.Unmarshal(msg, &m)
-
-	d, err := lib.Decode(m.Sender)
-	if err != nil {
-		return m, lib.ErrEncodingFault
-	}
-	m.Sender = string(d)
-
-	d, err = lib.Decode(m.MessageId)
-	if err != nil {
-		return m, lib.ErrEncodingFault
-	}
-	m.MessageId = string(d)
-
-	d, err = lib.Decode(m.Signature)
-	if err != nil {
-		return m, lib.ErrEncodingFault
-	}
-	m.Signature = string(d)
-
-	//Marshall automatically encode Json.rawMessage in Base64
-	d, err = lib.Decode(string(m.Data))
-	if err != nil {
-		return m, lib.ErrEncodingFault
-	}
-	m.Data = d
-
-	for i := 0; i < len(m.WitnessSignatures); i++ {
-		d, err = lib.Decode(m.WitnessSignatures[i])
-		if err != nil {
-			return m, lib.ErrEncodingFault
-		}
-		m.WitnessSignatures[i] = string(d)
-	}
 	return m, err
 }
-func ParseWitnessSignature(witnessSignatures string) (message.ItemWitnessSignatures, error) {
-	m := message.ItemWitnessSignatures{}
-	err := json.Unmarshal([]byte(witnessSignatures), &m)
-
-	d, err := lib.Decode(m.Signature)
-	if err != nil {
-		return m, lib.ErrEncodingFault
-	}
-	m.Signature = string(d)
-
-	d, err = lib.Decode(m.Witness)
-	if err != nil {
-		return m, lib.ErrEncodingFault
-	}
-	m.Witness = string(d)
 
+// ParseWitnessSignature parses a json.RawMessage into a message.ItemWitnessSignatures structure.
+func ParseWitnessSignature(witnessSignatures json.RawMessage) (message.ItemWitnessSignatures, error) {
+	m := message.ItemWitnessSignatures{}
+	err := json.Unmarshal(witnessSignatures, &m)
 	return m, err
 }
+
+// ParseData parses a json.RawMessage into a message.Data structure. It calls dataConstAreValid to check restrictions
+// defined in the jsonRPC protocol.
+// Careful that message.Data is a map and as such fields must be used with m["field"] until it is being decoded in the specific structure (and then m.field)
 func ParseData(data string) (message.Data, error) {
 	m := message.Data{}
 	err := json.Unmarshal([]byte(data), &m)
-	return m, err
+	if dataConstAreValid(m) {
+		return m, err
+	} else {
+		return m, lib.ErrRequestDataInvalid
+	}
+}
+
+// dataConstAreValid verifies following constraints :
+// * object is one of "lao", "message", "meeting"
+// * action is one of "create", update_properties", "state", "witness"
+// * creation and last modified are positive integer
+func dataConstAreValid(m message.Data) bool {
+	// TODO need to update as it grows
```

Comment:
what do we do @Huntsekkern then ? check each possibility or only the name of the fields ?

Reply:
I think we can just remove the todo, and put a WARNING at the end of the function's description

Reply:
done